### PR TITLE
HHL Slits Added to APS Tools

### LINKED ISSUE #923: Hoist support for HHL white beam slits from haven
HHL: High Heat Load

image

@canismarko wrote (on Teams):

> The apstools support might look like this ophyd device I wrote for haven that matches the interface for our blade slits:
https://github.com/spc-group/haven/blob/91a4e4f8b0c6f75f9ae8bf720412e75d634ef125/src/haven/instrument/slits.py#L66

### PR DESCRIPTION
**Pull request by MDecarabas**
- close #923

### CODE REVIEW

**Comment by prjemian** [comment]
> [action-black] Checking python code using the black formatter...
would reformat /github/workspace/apstools/devices/hhl_slits.py
> 
> Oh no! 💥 💔 💥
> 1 file would be reformatted, 159 files would be left unchanged.

**Comment by prjemian** [comment]
More files need to be updated:

https://github.com/BCDA-APS/apstools/blob/7201a73a65be24e8253fb4d73265c55c32075c3d/apstools/devices/__init__.py#L66

```suggestion

from .hhl_slits import HHLSlits
```

**Comment by MDecarabas** [comment]
> > [action-black] Checking python code using the black formatter...
> > would reformat /github/workspace/apstools/devices/hhl_slits.py
> > Oh no! 💥 💔 💥
> > 1 file would be reformatted, 159 files would be left unchanged.

No idea why this is happening. Formatted using vscode extension and then separately using the Conda-forge package. No change to the file

**Comment by prjemian** [comment]
https://github.com/BCDA-APS/apstools/blob/7201a73a65be24e8253fb4d73265c55c32075c3d/docs/source/api/_devices.rst?plain=1#L153

```suggestion
    ~apstools.devices.hhl_slits.HHLSlits
    ~apstools.devices.xia_slit.XiaSlit2D
```

https://github.com/BCDA-APS/apstools/blob/7201a73a65be24e8253fb4d73265c55c32075c3d/docs/source/api/_devices.rst?plain=1#L319

```suggestion

.. automodule:: apstools.devices.hhl_slits
    :members:
    :private-members:
    :show-inheritance:
    :inherited-members:```

**Comment by prjemian** [comment]
> > > [action-black] Checking python code using the black formatter...
> > > would reformat /github/workspace/apstools/devices/hhl_slits.py
> > > Oh no! 💥 💔 💥
> > > 1 file would be reformatted, 159 files would be left unchanged.
> 
> No idea why this is happening. Formatted using vscode extension and then separately using the Conda-forge package. No change to the file

I agree.  The logs are not informative.  I can check out your branch and diagnose locally.  Meanwhile, is there a verbosity level we  can add to the workflow?  https://github.com/BCDA-APS/apstools/blob/f4d15a962e858653a98c4521bc28d474385617b0/.github/workflows/code.yml#L39-L42

Consult the repo for the [action](https://github.com/rickstaa/action-black).

**Comment by canismarko** [comment]
Black v24 dropped recently, and picks up stuff that black v23 does not. I had to upgrade my local env to black v24.

**Comment by prjemian** [comment]
Locally, this is the change:
image

Blank line 45 removed after method signature.

**Comment by prjemian** [comment]
@MDecarabas - Did you test this with the actual EPICS controls, with the HHL slits at 25-ID?  That's one type of review.

File: apstools/devices/hhl_slits.py
Comment (prjemian) [inline apstools/devices/hhl_slits.py:9]: Could drop the text `OPHYD Device`.  That could apply to just about any code in this directory.
image

File: docs/source/api/_devices.rst
Comment (prjemian) [inline docs/source/api/_devices.rst:151]: Documentation will look like this:
image

The way to get documentation for the `FCpt` components is to add docstrings to each?  Not sure it's worth it now.  Inclined to skip it.

File: apstools/devices/hhl_slits.py
Comment (prjemian) [reply apstools/devices/hhl_slits.py:9]: This proposition is your call.  If you wish, change it before merging.

File: docs/source/api/_devices.rst
Comment (prjemian) [reply docs/source/api/_devices.rst:151]: Nothing to be changed here. Looks good!

**Review by prjemian — APPROVED** [review]
LGTM!

**Comment by prjemian** [comment]
Er, one more thing to update is the CHANGES.rst file (the release notes).

**Comment by prjemian** [comment]
This file: https://github.com/BCDA-APS/apstools/blob/hhl_slits/CHANGES.rst

Add entry for the new support, following previous examples.

**Comment by MDecarabas** [comment]
> @MDecarabas - Did you test this with the actual EPICS controls, with the HHL slits at 25-ID? That's one type of review.

Yes I did

**Comment by prjemian** [comment]
Change history addition, such as: https://github.com/BCDA-APS/apstools/blob/7a5eaca0a374d1a758d1e250dca8da3193c8836a/CHANGES.rst#L43-L47

(New section under the 1.6.19 release notes.)

**Comment by prjemian** [comment]
Also note the expected date of Feb 29 has been pushed by a month to Mar 31.  Feb 29 was not realistic.

**Comment by prjemian** [comment]
You can change that text, as well.

**Pull request merged**
